diff --git a/configure b/configure
index 0ed5de8409..224378e4c2 100755
--- a/configure
+++ b/configure
@@ -4352,10 +4352,10 @@ fi

 if enabled cuda_nvcc; then
     nvcc_default="nvcc"
-    nvccflags_default="-gencode arch=compute_30,code=sm_30 -O2"
+    nvccflags_default="-gencode arch=compute_75,code=sm_75 -O2"
 else
     nvcc_default="clang"
-    nvccflags_default="--cuda-gpu-arch=sm_30 -O2"
+    nvccflags_default="--cuda-gpu-arch=sm_75 -O2"
     NVCC_C=""
 fi

